README.md
# GPCRDB node for KNIME

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3257985.svg)](https://doi.org/10.5281/zenodo.3257985)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=3D-e-Chem_knime-gpcrdb&metric=alert_status)](https://sonarcloud.io/summary/new_code?id=3D-e-Chem_knime-gpcrdb)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=3D-e-Chem_knime-gpcrdb&metric=coverage)](https://sonarcloud.io/summary/new_code?id=3D-e-Chem_knime-gpcrdb)
[![Java CI with Maven](https://github.com/3D-e-Chem/knime-gpcrdb/actions/workflows/ci.yml/badge.svg)](https://github.com/3D-e-Chem/knime-gpcrdb/actions/workflows/ci.yml)

KNIME plugin for retrieving data from [https://gpcrdb.org](https://gpcrdb.org), GPCRdb website contains data, web tools and diagrams for G protein-coupled receptors (GPCRs).

## Installation

Requirements:

* KNIME, https://www.knime.org, version 5.1 or higher

Steps to get GPCRDB nodes inside KNIME:

1. Goto Help > Install new software ... menu
2. Press add button
3. Fill text fields with `https://3d-e-chem.github.io/updates/5.1`
4. Select --all sites-- in work with pulldown
5. Open KNIME 3D-e-Chem Contributions folder
6. Select GPCRDB
7. Install software & restart

## Usage

See example workflow in `examples` folder.

## Build

Requires Maven and JDK 8.

```shell
mvn verify
```

Jar has been made in `plugin/target` folder.
An Eclipse update site will be made in `p2/target/repository` repository.

## Development

Steps to get development environment setup based on https://github.com/knime/knime-sdk-setup#sdk-setup:

1. Install Java 17
2. Install Eclipse for [RCP and RAP developers](https://www.eclipse.org/downloads/packages/installer)
3. Configure Java 17 inside Eclipse Window > Preferences > Java > Installed JREs
4. Import this repo as an Existing Maven project
5. Activate target platform by going to Window > Preferences > Plug-in Development > Target Platform and check the `KNIME Analytics Platform (5.1) - nl.esciencecenter.e3dchem.knime.molviewer.targetplatform/KNIME-AP-5.1.target` target definition.

During import the Tycho Eclipse providers must be installed.

## New release

1. Update versions in pom files with `mvn org.eclipse.tycho:tycho-versions-plugin:set-version -DnewVersion=<version>-SNAPSHOT` command.
2. Commit and push changes
3. Create package with `mvn package`, will create update site in `p2/target/repository`
4. Test node by installing it from local update site
5. Append new release to 3D-e-Chem update site
    1. Make clone of [https://github.com/3D-e-Chem/3D-e-Chem.github.io](https://github.com/3D-e-Chem/3D-e-Chem.github.io) repo
    2. Append release to 3D-e-Chem update site with `mvn install -Dtarget.update.site=<3D-e-Chem repo/updates>`
6. Commit and push changes in this repo and 3D-e-Chem.github.io repo
7. Create a Github release
8. Update Zenodo entry
    1. Fix authors
    2. Fix license
9. Make nodes available to 3D-e-Chem KNIME feature by following steps at https://github.com/3D-e-Chem/knime-node-collection#new-release

## Create GPCRDB client

1. Download swagger code generator

  ```shell
  wget http://repo1.maven.org/maven2/io/swagger/swagger-codegen-cli/2.3.0/swagger-codegen-cli-2.3.0.jar
  ```

2. Download and unpack the swagger rewriter

3. Generate a Swagger spec for the client

  Install the swagger rewriter from https://github.com/3D-e-Chem/swagger-rewriter

  ```shell
  swagger-rewriter/bin/swagger-rewriter \
  https://gpcrdb.org/services/reference/api-docs/ \
  client-config/swagger-rewriter.config.yml \
  client-config/gpcrdb.swagger-spec.json
  ```

3.1 Optionally, make manual changes to client-config/gpcrdb.swagger-spec.json

4. Generate a client for GPCRDB web service using the rewritten spec

  ```shell
  java -jar swagger-codegen-cli-2.3.0.jar generate \
  --input-spec client-config/gpcrdb.swagger-spec.json \
  --output client \
  --lang java \
  --config client-config/swagger-codegen.config.json
  ```

5. Compile client

  ```shell
  cd client
  mvn package
  ```

6. Make client jar and it's dependencies available in plugin

  ```shell
  cp -r target/lib/* target/*jar ../plugin/lib/
  ```

7. Remove test dependencies

  ```shell
  rm plugin/lib/*-tests.jar plugin/lib/junit* plugin/lib/hamcrest*
  ```

8. Update `plugin/META-INF/MANIFEST.MF`, `plugin/build.properties` files to reflect contents of lib/

## Create stub recordings for integration tests

The test workflow are tested against a mocked web server and not the actual https://gpcrdb.org site.
The mock server is called [WireMock](http://WireMock.org/) and normally gives empty responses.
To have WireMock server return filled responses, stubs stored in `tests/src/test/resources/` directory must be provided.
The stubs can be recorded by starting a WireMock server in recording mode by:

```shell
java -jar tests/lib/wiremock-standalone-2.5.0.jar --proxy-all="https://gpcrdb.org/" \
--port=8089 --record-mappings --verbose --root-dir=tests/src/test/resources/
```

Then in a KNIME workflow in the GPCRDB nodes set the base path to http://localhost:8089.
Executing the workflow will fetch data from https://gpcrdb.org/ via the WireMock server and cause new stubs to be recorded in the `tests/src/test/resources/` directory.

To run the test workflows from inside KNIME desktop environment start the WireMock server in mock mode by:

```shell
java -jar tests/lib/wiremock-standalone-2.5.0.jar --port=8089 --verbose --root-dir=tests/src/test/resources/
```

Then import the test workflows in `tests/src/knime/` directory, select the workflow in the KNIME explorer and in the context menu (right-click) select `Run as workflow test`.

## References

* V Isberg, S Mordalski, C Munk, K Rataj, K Harpsøe, AS Hauser, B Vroling, AJ Bojarski, G Vriend, DE Gloriam. “GPCRdb: an information system for G protein-coupled receptors”, 2016, Nucleic Acids Res., 44, D356-D364. [10.1093/nar/gkv1178](http://dx.doi.org/10.1093/nar/gkv1178)
* V Isberg, B Vroling, R van der Kant, K Li, G Vriend* and DE Gloriam*, “GPCRDB: an information system for G protein-coupled receptors”, 2014, Nucleic Acids Res., 42 (D1), D422-D425. [10.1093/nar/gkv1178](http://dx.doi.org/10.1093/nar/gkv1178)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
Formatted as described on http://keepachangelog.com/.

## [Unreleased]

## [1.4.3] - 2023-11-21

### Changes

- Requires KNIME 5.1

## [1.4.2] - 2020-04-06

Nodes in existing workflows giving not found errors should be replaced.

### Fixed

- Not found error [#32](https://github.com/3D-e-Chem/knime-gpcrdb/issues/320

## [1.4.1] - 2019-06-27

### Changed

- Compatible with KNIME 4 [#30](https://github.com/3D-e-Chem/knime-gpcrdb/issues/30)

## [1.4.0] - 2018-04-04

### Fixed

- Add all fields that are returned by mutant ws endpoint to KNIME node (#29)

## [1.3.0] - 2017-09-01

### Fixed

- Mutant api changed (#29)

## [1.2.0] - 2017-03-01

### Added

- Node to fetch all GPCRDB structures (#27)

### Fixed

- Use rowkey generator instead of custom rowkey (#28)

## [1.1.0] - 2017-01-12

### Added

- Test coverage (#25)
- Timeout option (#21)
- Workflow test for all nodes (#24)

### Changed

- Nicer message for Timeout and not found exceptions (#23)

### Fixed

- Change default port names to specific names (#22)

## [1.0.15] - 2016-10-18

### Fixed

- Protein with single structure produce error (#8)

## [1.0.14] - 2016-10-18

### Added

- Test workflow using WireMock server (#17)

### Fixed

- Protein with single structure produce error (#8)

## [1.0.13] - 2016-09-13

### Changed

- Client no longer needs to be modified after generation, now the Swagger spec is adjusted before generation (#16)

### Fixed

- Based client on okhttp+gson library, a non jaxrs based implementation (#16)

## [1.0.12] - 2016-07-18

### Changed

- Moved nodes under /community/3d-e-chem (#15)

## [1.0.11] - 2016-06-07

### Added

- Additional input and node configuration checks in line with KNIME specification

## [1.0.10] - 2016-05-20

### Changed

- Shortened maven modules names

### Fixed

- Optional publication when requesting structures of protein (#12)

## [1.0.9] - 2016-05-13

### Fixed

- Protein with single structure produce error (#8)

### Changed

- Input is case-insensitive (#10)

## [1.0.8] - 2016-04-18

### Added

- Node for retrieving protein similarity (#5)

## [1.0.7] - 2016-03-29

### Added

- Node for retrieving mutations of protein (#4)
- Node for retrieving interactions of structure with its ligands (#6)

## [1.0.6] - 2016-01-27

### Changed

- Return alternative numbers as map (#7)


plugin/src/nl/esciencecenter/e3dchem/gpcrdb/mutants/MutantsNodeModel.java
package nl.esciencecenter.e3dchem.gpcrdb.mutants;

import java.io.File;
import java.io.IOException;
import java.util.List;

import org.knime.core.data.DataCell;
import org.knime.core.data.DataColumnSpec;
import org.knime.core.data.DataColumnSpecCreator;
import org.knime.core.data.DataRow;
import org.knime.core.data.DataTableSpec;
import org.knime.core.data.RowKey;
import org.knime.core.data.def.DefaultRow;
import org.knime.core.data.def.DoubleCell;
import org.knime.core.data.def.IntCell;
import org.knime.core.data.def.StringCell;
import org.knime.core.node.BufferedDataContainer;
import org.knime.core.node.BufferedDataTable;
import org.knime.core.node.CanceledExecutionException;
import org.knime.core.node.ExecutionContext;
import org.knime.core.node.ExecutionMonitor;
import org.knime.core.node.InvalidSettingsException;
import org.knime.core.node.NodeSettingsRO;
import org.knime.core.node.NodeSettingsWO;
import org.knime.core.node.defaultnodesettings.SettingsModelString;

import nl.esciencecenter.e3dchem.gpcrdb.GpcrdbNodeModel;
import nl.esciencecenter.e3dchem.gpcrdb.client.ApiException;
import nl.esciencecenter.e3dchem.gpcrdb.client.ServicesmutantsApi;
import nl.esciencecenter.e3dchem.gpcrdb.client.model.MutationSerializer;

/**
 * This is the model implementation of Mutants. Get a list of mutants of a protein
 *
 */
public class MutantsNodeModel extends GpcrdbNodeModel {

    public static final String CFGKEY_INPUTCOLUMNNAME = "In Column";

    private final SettingsModelString m_inputColumnName = new SettingsModelString(CFGKEY_INPUTCOLUMNNAME, null);

    private ServicesmutantsApi service = new ServicesmutantsApi(getApiClient());

    /**
     * Constructor for the node model.
     */
    public MutantsNodeModel() {

        // TODO one incoming port and one outgoing port is assumed
        super(1, 1);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected BufferedDataTable[] execute(final BufferedDataTable[] inData, final ExecutionContext exec) throws Exception {
        // the data table spec of the single output table,
        // the table will have three columns:
        DataColumnSpec[] allColSpecs = new DataColumnSpec[23];
        allColSpecs[0] = new DataColumnSpecCreator("Entry name", StringCell.TYPE).createSpec();
        allColSpecs[1] = new DataColumnSpecCreator("Sequence number", IntCell.TYPE).createSpec();
        allColSpecs[2] = new DataColumnSpecCreator("Mutation from", StringCell.TYPE).createSpec();
        allColSpecs[3] = new DataColumnSpecCreator("Mutation to", StringCell.TYPE).createSpec();
        allColSpecs[4] = new DataColumnSpecCreator("Ligand name", StringCell.TYPE).createSpec();
        allColSpecs[5] = new DataColumnSpecCreator("Ligand ID type", StringCell.TYPE).createSpec();
        allColSpecs[6] = new DataColumnSpecCreator("Ligand ID", StringCell.TYPE).createSpec();
        allColSpecs[7] = new DataColumnSpecCreator("Ligand Class", StringCell.TYPE).createSpec();
        allColSpecs[8] = new DataColumnSpecCreator("Measure type", StringCell.TYPE).createSpec();
        allColSpecs[9] = new DataColumnSpecCreator("Assay type", StringCell.TYPE).createSpec();
        allColSpecs[10] = new DataColumnSpecCreator("WT activity/affinity", DoubleCell.TYPE).createSpec();
        allColSpecs[11] = new DataColumnSpecCreator("Unit", StringCell.TYPE).createSpec();
        allColSpecs[12] = new DataColumnSpecCreator("Quantifier symbol", StringCell.TYPE).createSpec();
        allColSpecs[13] = new DataColumnSpecCreator("Quantitative", DoubleCell.TYPE).createSpec();
        allColSpecs[14] = new DataColumnSpecCreator("Qualitative", StringCell.TYPE).createSpec();
        allColSpecs[15] = new DataColumnSpecCreator("Receptor property", StringCell.TYPE).createSpec();
        allColSpecs[16] = new DataColumnSpecCreator("WT data", DoubleCell.TYPE).createSpec();
        allColSpecs[17] = new DataColumnSpecCreator("Quantitative symbol", StringCell.TYPE).createSpec();
        allColSpecs[18] = new DataColumnSpecCreator("Mutant effect", DoubleCell.TYPE).createSpec();
        allColSpecs[19] = new DataColumnSpecCreator("Mutant effect (%)", DoubleCell.TYPE).createSpec();
        allColSpecs[20] = new DataColumnSpecCreator("Qualitative effect", StringCell.TYPE).createSpec();
        allColSpecs[21] = new DataColumnSpecCreator("Agonist name", StringCell.TYPE).createSpec();
        allColSpecs[22] = new DataColumnSpecCreator("Reference", StringCell.TYPE).createSpec();

        // TODO add foldchange 
        DataTableSpec outputSpec = new DataTableSpec(allColSpecs);

        // the execution context will provide us with storage capacity, in this
        // case a data container to which we will add rows sequentially
        // Note, this container can also handle arbitrary big data tables, it
        // will buffer to disc if necessary.
        BufferedDataContainer container = exec.createDataContainer(outputSpec);

        BufferedDataTable table = inData[0];
        long rowCount = table.size();
        long currentRow = 0;
        int columnIndex = table.getDataTableSpec().findColumnIndex(m_inputColumnName.getStringValue());

        for (DataRow inrow : table) {
            String entryName = ((StringCell) inrow.getCell(columnIndex)).getStringValue();

            fetchMutants(entryName, container);

            // check if the user cancelled the execution
            exec.checkCanceled();
            // report progress
            exec.setProgress((double) currentRow / rowCount, " processing row " + currentRow);
            currentRow++;
        }

        // once we are done, we close the container and return its table
        container.close();
        BufferedDataTable out = container.getTable();
        return new BufferedDataTable[] { out };
    }

    public void fetchMutants(String entryName, BufferedDataContainer container) throws ApiException {
    	entryName = entryName.toLowerCase();
        List<MutationSerializer> mutants = service.mutantListGET(entryName);
        long currentRow = 0;
        for (MutationSerializer mutant : mutants) {
            RowKey key = new RowKey(entryName + " - " + currentRow);
            // the cells of the current row, the types of the cells must
            // match
            // the column spec (see above)
            DataCell[] cells = new DataCell[23];
            cells[0] = new StringCell(mutant.getProtein());
            cells[1] = new IntCell(mutant.getMutationPos());
            cells[2] = new StringCell(mutant.getMutationFrom());
            cells[3] = new StringCell(mutant.getMutationTo());
            cells[4] = new StringCell(mutant.getLigandName());
            cells[5] = new StringCell(mutant.getLigandIdtype());
            cells[6] = new StringCell(mutant.getLigandId());
            cells[7] = new StringCell(mutant.getLigandClass());
            cells[8] = new StringCell(mutant.getExpType());
            cells[9] = new StringCell(mutant.getExpFunc());
            cells[10] = new DoubleCell(mutant.getExpWtValue());
            cells[11] = new StringCell(mutant.getExpWtUnit());
            cells[12] = new StringCell(mutant.getExpMuEffectSign());
            cells[13] = new DoubleCell(mutant.getExpMuEffectValue());
            cells[14] = new StringCell(mutant.getExpMuEffectQual());
            cells[15] = new StringCell(mutant.getOptType());
            cells[16] = new DoubleCell(mutant.getOptWt());
            cells[17] = new StringCell(mutant.getOptSign());
            cells[18] = new DoubleCell(mutant.getOptMu());
            cells[19] = new DoubleCell(mutant.getOptPercentage());
            cells[20] = new StringCell(mutant.getOptQual());
            cells[21] = new StringCell(mutant.getOptAgonist());
            cells[22] = new StringCell(mutant.getReference());
            DataRow row = new DefaultRow(key, cells);
            container.addRowToTable(row);
            currentRow++;
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void reset() {
        // TODO Code executed on reset.
        // Models build during execute are cleared here.
        // Also data handled in load/saveInternals will be erased here.
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected DataTableSpec[] configure(final DataTableSpec[] inSpecs) throws InvalidSettingsException {
    	if (inSpecs.length > 0 && inSpecs[0] != null){
        	int columnIndex = inSpecs[0].findColumnIndex(m_inputColumnName.getStringValue());
        	if (columnIndex < 0){
        		throw new InvalidSettingsException("No valid input column selected");
        	}
        }    	
    	
        return new DataTableSpec[] { null };
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void saveSettingsTo(final NodeSettingsWO settings) {

        super.saveSettingsTo(settings);
        m_inputColumnName.saveSettingsTo(settings);

    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void loadValidatedSettingsFrom(final NodeSettingsRO settings) throws InvalidSettingsException {

        // TODO load (valid) settings from the config object.
        // It can be safely assumed that the settings are valided by the
        // method below.

        super.loadValidatedSettingsFrom(settings);
        m_inputColumnName.loadSettingsFrom(settings);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void validateSettings(final NodeSettingsRO settings) throws InvalidSettingsException {

        // TODO check if the settings could be applied to our model
        // e.g. if the count is in a certain range (which is ensured by the
        // SettingsModel).
        // Do not actually set any values of any member variables.

        super.validateSettings(settings);
        m_inputColumnName.validateSettings(settings);

    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void loadInternals(final File internDir, final ExecutionMonitor exec)
            throws IOException, CanceledExecutionException {

        // TODO load internal data.
        // Everything handed to output ports is loaded automatically (data
        // returned by the execute method, models loaded in loadModelContent,
        // and user settings set through loadSettingsFrom - is all taken care
        // of). Load here only the other internals that need to be restored
        // (e.g. data used by the views).

    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected void saveInternals(final File internDir, final ExecutionMonitor exec)
            throws IOException, CanceledExecutionException {

        // TODO save internal models.
        // Everything written to output ports is saved automatically (data
        // returned by the execute method, models saved in the saveModelContent,
        // and user settings saved through saveSettingsTo - is all taken care
        // of). Save here only the other internals that need to be preserved
        // (e.g. data used by the views).

    }

    public ServicesmutantsApi getService() {
        return service;
    }

    public void setService(ServicesmutantsApi service) {
        this.service = service;
    }

}
